### PR TITLE
T5169: Add PoC for generating CGNAT rules rfc6888

### PR DESCRIPTION
**Pull request by sever-sever**
## Change Summary

Add PoC to generate CGNAT rules (that's why classes and templates are in the same file now).

https://datatracker.ietf.org/doc/html/rfc6888
Not all requirements are implemented, but some of them. Implemented:

REQ-2
```
A CGN MUST have a default "IP address pooling" behavior of "Paired"
CGN must use the same external IP
      address mapping for all sessions associated with the same internal
      IP address, be they TCP, UDP, ICMP, something else, or a mix of
      different protocols.
```

REQ-3
```
The CGN function SHOULD NOT have any limitations on the size
      or the contiguity of the external address pool
```

REQ-4
```
A CGN MUST support limiting the number of external ports (or,
      equivalently, "identifiers" for ICMP) that are assigned per
      subscriber
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

 * https://vyos.dev/T5169

## Component(s) name

cgnat, nat
## How to test
Example with prefixes:
```
set nat cgnat pool external ext1 external-port-range '1024-65535'
set nat cgnat pool external ext1 per-user-limit port '1000'
set nat cgnat pool external ext1 range 192.0.2.222/32
set nat cgnat pool internal int1 range '100.64.0.0/28'
set nat cgnat rule 10 source pool 'int1'
set nat cgnat rule 10 translation pool 'ext1'

```
Check nftables:
```
vyos@r4# sudo nft -s list table ip cgnat
table ip cgnat {
	map tcp_nat_map {
		type ipv4_addr : interval ipv4_addr . inet_service
		flags interval
		elements = { 100.64.0.0 : 192.0.2.222 . 1024-2023, 100.64.0.1 : 192.0.2.222 . 2024-3023,
			     100.64.0.2 : 192.0.2.222 . 3024-4023, 100.64.0.3 : 192.0.2.222 . 4024-5023,
			     100.64.0.4 : 192.0.2.222 . 5024-6023, 100.64.0.5 : 192.0.2.222 . 6024-7023,
			     100.64.0.6 : 192.0.2.222 . 7024-8023, 100.64.0.7 : 192.0.2.222 . 8024-9023,
			     100.64.0.8 : 192.0.2.222 . 9024-10023, 100.64.0.9 : 192.0.2.222 . 10024-11023,
			     100.64.0.10 : 192.0.2.222 . 11024-12023, 100.64.0.11 : 192.0.2.222 . 12024-13023,
			     100.64.0.12 : 192.0.2.222 . 13024-14023, 100.64.0.13 : 192.0.2.222 . 14024-15023,
			     100.64.0.14 : 192.0.2.222 . 15024-16023, 100.64.0.15 : 192.0.2.222 . 16024-17023 }
	}

	map udp_nat_map {
		type ipv4_addr : interval ipv4_addr . inet_service
		flags interval
		elements = { 100.64.0.0 : 192.0.2.222 . 1024-2023, 100.64.0.1 : 192.0.2.222 . 2024-3023,
			     100.64.0.2 : 192.0.2.222 . 3024-4023, 100.64.0.3 : 192.0.2.222 . 4024-5023,
			     100.64.0.4 : 192.0.2.222 . 5024-6023, 100.64.0.5 : 192.0.2.222 . 6024-7023,
			     100.64.0.6 : 192.0.2.222 . 7024-8023, 100.64.0.7 : 192.0.2.222 . 8024-9023,
			     100.64.0.8 : 192.0.2.222 . 9024-10023, 100.64.0.9 : 192.0.2.222 . 10024-11023,
			     100.64.0.10 : 192.0.2.222 . 11024-12023, 100.64.0.11 : 192.0.2.222 . 12024-13023,
			     100.64.0.12 : 192.0.2.222 . 13024-14023, 100.64.0.13 : 192.0.2.222 . 14024-15023,
			     100.64.0.14 : 192.0.2.222 . 15024-16023, 100.64.0.15 : 192.0.2.222 . 16024-17023 }
	}

	map icmp_nat_map {
		type ipv4_addr : interval ipv4_addr . inet_service
		flags interval
		elements = { 100.64.0.0 : 192.0.2.222 . 1024-2023, 100.64.0.1 : 192.0.2.222 . 2024-3023,
			     100.64.0.2 : 192.0.2.222 . 3024-4023, 100.64.0.3 : 192.0.2.222 . 4024-5023,
			     100.64.0.4 : 192.0.2.222 . 5024-6023, 100.64.0.5 : 192.0.2.222 . 6024-7023,
			     100.64.0.6 : 192.0.2.222 . 7024-8023, 100.64.0.7 : 192.0.2.222 . 8024-9023,
			     100.64.0.8 : 192.0.2.222 . 9024-10023, 100.64.0.9 : 192.0.2.222 . 10024-11023,
			     100.64.0.10 : 192.0.2.222 . 11024-12023, 100.64.0.11 : 192.0.2.222 . 12024-13023,
			     100.64.0.12 : 192.0.2.222 . 13024-14023, 100.64.0.13 : 192.0.2.222 . 14024-15023,
			     100.64.0.14 : 192.0.2.222 . 15024-16023, 100.64.0.15 : 192.0.2.222 . 16024-17023 }
	}

	map other_nat_map {
		type ipv4_addr : interval ipv4_addr
		flags interval
		elements = { 100.64.0.0 : 192.0.2.222/32, 100.64.0.1 : 192.0.2.222/32,
			     100.64.0.2 : 192.0.2.222/32, 100.64.0.3 : 192.0.2.222/32,
			     100.64.0.4 : 192.0.2.222/32, 100.64.0.5 : 192.0.2.222/32,
			     100.64.0.6 : 192.0.2.222/32, 100.64.0.7 : 192.0.2.222/32,
			     100.64.0.8 : 192.0.2.222/32, 100.64.0.9 : 192.0.2.222/32,
			     100.64.0.10 : 192.0.2.222/32, 100.64.0.11 : 192.0.2.222/32,
			     100.64.0.12 : 192.0.2.222/32, 100.64.0.13 : 192.0.2.222/32,
			     100.64.0.14 : 192.0.2.222/32, 100.64.0.15 : 192.0.2.222/32 }
	}

	chain POSTROUTING {
		type nat hook postrouting priority srcnat; policy accept;
		ip protocol tcp counter snat ip to ip saddr map @tcp_nat_map
		ip protocol udp counter snat ip to ip saddr map @udp_nat_map
		ip protocol icmp counter snat ip to ip saddr map @icmp_nat_map
		counter snat ip to ip saddr map @other_nat_map
	}
}

```
Example with ranges:
```
set nat cgnat pool external ext1 external-port-range '1024-65535'
set nat cgnat pool external ext1 per-user-limit port '8000'
set nat cgnat pool external ext1 range 192.0.2.1-192.0.2.2
set nat cgnat pool internal int1 range '100.64.0.1-100.64.0.16'
set nat cgnat rule 10 source pool 'int1'
set nat cgnat rule 10 translation pool 'ext1'
```
Check nft for ranges:
```
vyos@r4# sudo nft -s list table ip cgnat
table ip cgnat {
	map tcp_nat_map {
		type ipv4_addr : interval ipv4_addr . inet_service
		flags interval
		elements = { 100.64.0.1 : 192.0.2.1 . 1024-9023, 100.64.0.2 : 192.0.2.1 . 9024-17023,
			     100.64.0.3 : 192.0.2.1 . 17024-25023, 100.64.0.4 : 192.0.2.1 . 25024-33023,
			     100.64.0.5 : 192.0.2.1 . 33024-41023, 100.64.0.6 : 192.0.2.1 . 41024-49023,
			     100.64.0.7 : 192.0.2.1 . 49024-57023, 100.64.0.8 : 192.0.2.1 . 57024-65023,
			     100.64.0.9 : 192.0.2.2 . 1024-9023, 100.64.0.10 : 192.0.2.2 . 9024-17023,
			     100.64.0.11 : 192.0.2.2 . 17024-25023, 100.64.0.12 : 192.0.2.2 . 25024-33023,
			     100.64.0.13 : 192.0.2.2 . 33024-41023, 100.64.0.14 : 192.0.2.2 . 41024-49023,
			     100.64.0.15 : 192.0.2.2 . 49024-57023, 100.64.0.16 : 192.0.2.2 . 57024-65023 }
	}

	map udp_nat_map {
		type ipv4_addr : interval ipv4_addr . inet_service
		flags interval
		elements = { 100.64.0.1 : 192.0.2.1 . 1024-9023, 100.64.0.2 : 192.0.2.1 . 9024-17023,
			     100.64.0.3 : 192.0.2.1 . 17024-25023, 100.64.0.4 : 192.0.2.1 . 25024-33023,
			     100.64.0.5 : 192.0.2.1 . 33024-41023, 100.64.0.6 : 192.0.2.1 . 41024-49023,
			     100.64.0.7 : 192.0.2.1 . 49024-57023, 100.64.0.8 : 192.0.2.1 . 57024-65023,
			     100.64.0.9 : 192.0.2.2 . 1024-9023, 100.64.0.10 : 192.0.2.2 . 9024-17023,
			     100.64.0.11 : 192.0.2.2 . 17024-25023, 100.64.0.12 : 192.0.2.2 . 25024-33023,
			     100.64.0.13 : 192.0.2.2 . 33024-41023, 100.64.0.14 : 192.0.2.2 . 41024-49023,
			     100.64.0.15 : 192.0.2.2 . 49024-57023, 100.64.0.16 : 192.0.2.2 . 57024-65023 }
	}

	map icmp_nat_map {
		type ipv4_addr : interval ipv4_addr . inet_service
		flags interval
		elements = { 100.64.0.1 : 192.0.2.1 . 1024-9023, 100.64.0.2 : 192.0.2.1 . 9024-17023,
			     100.64.0.3 : 192.0.2.1 . 17024-25023, 100.64.0.4 : 192.0.2.1 . 25024-33023,
			     100.64.0.5 : 192.0.2.1 . 33024-41023, 100.64.0.6 : 192.0.2.1 . 41024-49023,
			     100.64.0.7 : 192.0.2.1 . 49024-57023, 100.64.0.8 : 192.0.2.1 . 57024-65023,
			     100.64.0.9 : 192.0.2.2 . 1024-9023, 100.64.0.10 : 192.0.2.2 . 9024-17023,
			     100.64.0.11 : 192.0.2.2 . 17024-25023, 100.64.0.12 : 192.0.2.2 . 25024-33023,
			     100.64.0.13 : 192.0.2.2 . 33024-41023, 100.64.0.14 : 192.0.2.2 . 41024-49023,
			     100.64.0.15 : 192.0.2.2 . 49024-57023, 100.64.0.16 : 192.0.2.2 . 57024-65023 }
	}

	map other_nat_map {
		type ipv4_addr : interval ipv4_addr
		flags interval
		elements = { 100.64.0.1 : 192.0.2.1/32, 100.64.0.2 : 192.0.2.1/32,
			     100.64.0.3 : 192.0.2.1/32, 100.64.0.4 : 192.0.2.1/32,
			     100.64.0.5 : 192.0.2.1/32, 100.64.0.6 : 192.0.2.1/32,
			     100.64.0.7 : 192.0.2.1/32, 100.64.0.8 : 192.0.2.1/32,
			     100.64.0.9 : 192.0.2.2/32, 100.64.0.10 : 192.0.2.2/32,
			     100.64.0.11 : 192.0.2.2/32, 100.64.0.12 : 192.0.2.2/32,
			     100.64.0.13 : 192.0.2.2/32, 100.64.0.14 : 192.0.2.2/32,
			     100.64.0.15 : 192.0.2.2/32, 100.64.0.16 : 192.0.2.2/32 }
	}

	chain POSTROUTING {
		type nat hook postrouting priority srcnat; policy accept;
		ip protocol tcp counter snat ip to ip saddr map @tcp_nat_map
		ip protocol udp counter snat ip to ip saddr map @udp_nat_map
		ip protocol icmp counter snat ip to ip saddr map @icmp_nat_map
		counter snat ip to ip saddr map @other_nat_map
	}
}

```
Try out of the range (ports in config more than available external-addresses/ports):
```
vyos@r4# set nat cgnat pool external ext1 per-user-limit port '16000'
[edit]
vyos@r4# commit
[ cgnat ]
Not enough ports available for the specified parameters!

[[cgnat]] failed
Commit failed
[edit]
vyos@r4# 

```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Comment by sarthurdev** [comment]
Feel like it should possibly fall under the main nat node?

`set nat cgnat ...`

**Comment by sever-sever** [comment]
> Feel like it should possibly fall under the main nat node?
> 
> `set nat cgnat ...`

@sarthurdev Exactly!! Fixed, thanks.

File: interface-definitions/nat_cgnat.xml.in [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        </properties>
+        <children>
+          <node name="pool">
+            <properties>
+              <help>External and internal pool parameters</help>
+            </properties>
+            <children>
+              <tagNode name="external">
+                <properties>
+                  <help>External pool name</help>
```
Comment (c-po) [inline interface-definitions/nat_cgnat.xml.in:18]: Please add constraint `constraint/alpha-numeric-hyphen-underscore.xml.i` here

File: interface-definitions/nat_cgnat.xml.in [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+                      </constraint>
+                    </properties>
+                    <defaultValue>1024-65535</defaultValue>
+                  </leafNode>
+                  <node name="per-user-limit">
+                    <properties>
+                      <help>Per user limits for the pool</help>
+                    </properties>
+                    <children>
+                      <leafNode name="port">
```
Comment (c-po) [inline interface-definitions/nat_cgnat.xml.in:47]: Please use `port-number.xml.i` building block

File: interface-definitions/nat_cgnat.xml.in [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                      <constraint>
+                        <validator name="ipv4-prefix"/>
+                        <validator name="ipv4-host"/>
+                        <validator name="ipv4-range"/>
+                      </constraint>
+                    </properties>
+                    <children>
+                      <leafNode name="seq">
+                        <properties>
+                          <help>Sequence</help>
```
Comment (c-po) [inline interface-definitions/nat_cgnat.xml.in:82]: Constraint? valueHelp?

File: interface-definitions/nat_cgnat.xml.in [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+                    </children>
+                  </tagNode>
+                </children>
+              </tagNode>
+              <tagNode name="internal">
+                <properties>
+                  <help>Internal pool settings</help>
+                </properties>
+                <children>
+                  <leafNode name="range">
```
Comment (c-po) [inline interface-definitions/nat_cgnat.xml.in:110]: Can we combine external and internal range to one XML include block?

File: interface-definitions/nat_cgnat.xml.in [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                        <validator name="ipv4-range"/>
+                      </constraint>
+                    </properties>
+                  </leafNode>
+                </children>
+              </tagNode>
+            </children>
+          </node>
+          <tagNode name="rule">
+            <properties>
```
Comment (c-po) [inline interface-definitions/nat_cgnat.xml.in:133]: Constraint? ValueHelp?

File: interface-definitions/nat_cgnat.xml.in [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+                </children>
+              </node>
+              <node name="translation">
+                <properties>
+                  <help>Translation parameters</help>
+                </properties>
+                <children>
+                  <leafNode name="pool">
+                    <properties>
+                      <help>Translation external pool</help>
```
Comment (c-po) [inline interface-definitions/nat_cgnat.xml.in:175]: Can we get the external pool completion helper?

File: src/conf_mode/nat_cgnat.py [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+from vyos.config import Config
+from vyos.utils.process import call
+from vyos import ConfigError
+from vyos import airbag
+airbag.enable()
+
+
+output_filename = '/run/nftables-cgnat.nft'
+
+BATCH_FILE_MAP_TEMPLATE = '''\
```
Comment (c-po) [inline src/conf_mode/nat_cgnat.py]: Any reason to not use the template directory under /data? Then J2lint can lint the template.

https://github.com/vyos/vyos-1x/tree/current/data/templates/firewall

**Review by c-po — CHANGES_REQUESTED** [review]

**Comment by c-po** [comment]
I also think `set nat cgnat` makes more sense though

File: interface-definitions/nat_cgnat.xml.in (hunk 4 above)
Comment (sever-sever) [reply interface-definitions/nat_cgnat.xml.in:110]: Probably not
The external pool should use sequences for the range to add priority to external address allocation.
The internal pool does not have sequences. However, we can add sequences to the internal pool if we want.
```
set nat cgnat pool external ext1 range 192.0.2.222/32 seq 1
set nat cgnat pool external ext1 range 192.0.2.253/32 seq 2
```
But internal is a single pool
```
set nat cgnat pool internal int1 range '100.64.0.0/28'
```

File: src/conf_mode/nat_cgnat.py (hunk 7 above)
Comment (sever-sever) [reply src/conf_mode/nat_cgnat.py]: If the PoC is OK, I will do it.

File: interface-definitions/nat_cgnat.xml.in (hunk 2 above)
Comment (sever-sever) [reply interface-definitions/nat_cgnat.xml.in:47]: The `port-number.xml.i` uses general help `Port number used by connection`.
It is completely incorrect, as we set the number of ports that will be allocated per user.
It is the reason for using a new leafNode port

File: interface-definitions/nat_cgnat.xml.in (hunk 1 above)
Comment (sever-sever) [reply interface-definitions/nat_cgnat.xml.in:18]: Done!

File: interface-definitions/nat_cgnat.xml.in (hunk 3 above)
Comment (sever-sever) [reply interface-definitions/nat_cgnat.xml.in:82]: Fixed!

File: interface-definitions/nat_cgnat.xml.in (hunk 5 above)
Comment (sever-sever) [reply interface-definitions/nat_cgnat.xml.in:133]: Done!

File: interface-definitions/nat_cgnat.xml.in (hunk 6 above)
Comment (sever-sever) [reply interface-definitions/nat_cgnat.xml.in:175]: Done!

**Comment by sever-sever** [comment]
> I also think `set nat cgnat` makes more sense though

It has the correct format now. The `cgnat` node under the `nat` section

File: src/conf_mode/nat_cgnat.py (hunk 7 above)
Comment (sever-sever) [reply src/conf_mode/nat_cgnat.py]: Done! 
Added Jinja2 template `nftables-cgnat.j2`

File: src/conf_mode/nat_cgnat.py [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+    tmp = run(f'nft --check --file {nftables_cgnat_config}')
+    if tmp > 0:
+        raise ConfigError('Configuration file errors encountered!')
+
+
+def apply(config):
+    if not config:
+        # Cleanup cgnat
+        cmd('nft delete table ip cgnat')
+        os.unlink(nftables_cgnat_config)
```
Comment (c-po) [inline src/conf_mode/nat_cgnat.py]: I'd add a condition to check if the file exists. What when a user deletes it for fun? An exception would emerge.

**Review by c-po — CHANGES_REQUESTED** [review]

File: src/conf_mode/nat_cgnat.py (hunk 8 above)
Comment (sever-sever) [reply src/conf_mode/nat_cgnat.py]: Done!

**Comment by fischerdouglas** [comment]
Is there anywhere on the VyOS horizon for BPA - Bulk Port Allocation (AKA LSN)?

**Comment by sever-sever** [comment]
@fischerdouglas Add a feature request on https://vyos.dev/ with some examples of what do you want.
The root task for CGN https://vyos.dev/T5169

**Pull request merged**